hypothesis-python/docs/tutorial/adapting-strategies.rst

```diff
@@ -0,0 +1,135 @@
+    def test_remainder_magnitude(a, b):
+        assume(b != 0)
+        # b will be nonzero here
+        assert abs(a % b) < abs(b)
+
+|assume| vs |strategy.filter|
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Where possible, you should use |strategy.filter|. Hypothesis can often rewrite simple filters into more efficient sampling methods than rejection sampling, and will retry filters several times instead of aborting the entire test case (as with |assume|).
+
+For more complex relationships that can't be expressed with |strategy.filter|, use |assume|.
+
+Here's an example of a test where we want to filter out two different types of examples:
+
+.. code-block:: python
+
+    from hypothesis import assume, given, strategies as st
+
+    @given(st.integers(), st.integers())
+    def test_floor_division_lossless_when_b_divides_a(a, b):
+        # we want to assume that:
+        # * b is nonzero, and
+        # * b divides a
+        assert (a // b) * b == a
+
+We could start by using |assume| for both:
+
+.. code-block:: python
+
+    from hypothesis import assume, given, strategies as st
+
+    @given(st.integers(), st.integers())
+    def test_floor_division_lossless_when_b_divides_a(a, b):
+        assume(b != 0)
+        assume(a % b == 0)
+        assert (a // b) * b == a
+
+And then notice that the ``b != 0`` condition can be moved into the strategy definition as a |strategy.filter| call:
+
+.. code-block:: python
+
+    from hypothesis import assume, given, strategies as st
+
+    @given(st.integers(), st.integers().filter(lambda n: n != 0))
+    def test_floor_division_lossless_when_b_divides_a(a, b):
+        assume(a % b == 0)
+        assert (a // b) * b == a
+
+However, the ``a % b == 0`` condition has to stay as an |assume|, because it expresses a more complicated relationship between ``a`` and ``b``.
+
+|assume| vs early-returning
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+One other way we could have avoided the divide-by-zero error inside the test function is to early-return when ``b == 0``:
+
+.. code-block:: python
+
+    from hypothesis import assume, given, strategies as st
+
+    @given(st.integers(), st.integers())
+    def test_remainder_magnitude(a, b):
+        if b == 0:
+            return
+        assert abs(a % b) < abs(b)
+
+While this would have avoided the divide-by-zero, early-returning is not the same as using |assume|. With |assume|, Hypothesis knows that a test case has been filtered out, and will not count it towards the |max_examples| limit. In contrast, early-returns are counted as a valid example. In more complicted cases, this could end up testing your code less than you expect, because many test cases get discarded without Hypothesis knowing about it.
+
+In addition, |assume| lets you skip the test case at any point in the test, even inside arbitrarily deep nestings of functions.
+
+You should always use |assume| rather than early-returning. |assume| is more idiomatic and allows Hypothesis more insight into your test.
```

Comment:
```suggestion
While this would have avoided the divide-by-zero, early-returning is not the same as using |assume|. With |assume|, Hypothesis knows that a test case has been filtered out, and will not count it towards the |max_examples| limit. In contrast, early-returns are counted as a passing test, even though the assertions didn't run!. In more complicted cases, this could end up testing your code less than you expect, because many test cases get discarded without Hypothesis knowing about it.

In addition, |assume| lets you skip the test case at any point in the test, even inside arbitrarily deep nestings of functions.
```